src/cockpit/channels/stream.py

```diff
@@ -54,6 +54,31 @@ def create_transport(self, loop: asyncio.AbstractEventLoop, options: Dict[str, o
         return SocketTransport(loop, self, connection)
 
 
+class TCPStreamChannel(ProtocolChannel):
+    payload = 'stream'
+    restrictions = (('port', None),)
+
+    def create_transport(self, loop: asyncio.AbstractEventLoop, options: Dict[str, object]) -> SocketTransport:
+        try:
+            port: int = int(options['port'])
+        except ValueError:
+            raise ChannelError('protocol-error', message='invalid "port" option for stream channel')
+        address = (options.get('address', 'localhost'), port)
+
+        try:
+            connection = socket.create_connection(address)
```

Comment:
This will block.

Reply:
Argh, yes -- let's use [asyncio's loop.create_connection()](https://docs.python.org/3/library/asyncio-eventloop.html?highlight=call%20soon%20threadsafe#opening-network-connections)

Reply:
@allisonkarlitskaya : I think I'll need your help here. `loop.create_connection()` (which seems to fit rather well here), as well as `asyncio.open_connection()` (which fits a bit less well, as we'd have to manually treat the reader and writer objects separately) are both coroutines. But all methods in `ProtocolChannel` are sync, i.e. I can't just put either of these calls into `create_transport()` or more generally into `do_open()`.

Should I use `AsyncChannel` for that? I think I'll see how to do the reading in `async def run()`, but not yet where writes enter the picture here.

Reply:
Take a look at https://docs.python.org/3/library/asyncio-eventloop.html#executing-code-in-thread-or-process-pools for a possible alternative.  This is what I'd do.

Reply:
That doesn't work for the same reason: One cannot call async functions in `do_open()` or `create_transport`. The asyncio and our Channel API don't seem to fit together very well.. I suppose I'll just start with a plain old thread, unless you have a better idea.

Reply:
I think we need to make `create_transport()` async, then.

Reply:
Reworked. This is actually kind of nice! I probably got some details wrong :grin: